gcs/constraints/circuit/circuit_scc.cc

```diff
@@ -0,0 +1,1250 @@
+#include <gcs/constraints/circuit/circuit.hh>
+#include <gcs/innards/propagators.hh>
+#include <iostream>
+#include <list>
+#include <random>
+#include <set>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+using std::cmp_equal;
+using std::cmp_less;
+using std::cmp_less_equal;
+using std::cmp_not_equal;
+using std::cout;
+using std::get;
+using std::holds_alternative;
+using std::list;
+using std::make_optional;
+using std::make_pair;
+using std::map;
+using std::min;
+using std::move;
+using std::mt19937;
+using std::nullopt;
+using std::optional;
+using std::pair;
+using std::random_device;
+using std::set;
+using std::string;
+using std::stringstream;
+using std::to_string;
+using std::tuple;
+using std::uniform_int_distribution;
+using std::unique_ptr;
+using std::variant;
+using std::vector;
+
+using namespace gcs;
+using namespace gcs::innards;
+
+namespace
+{
+
+    struct OrderingAssumption
+    {
+        ProofFlag assumption_flag;
+        long first;
+        long middle;
+        long last;
+    };
+
+    auto select_root(long n) -> long
+    {
+        // Might have better root selection in future
+        return 0;
+    }
+
+    struct SCCPropagatorData
+    {
+        // Data required for the modified Tarjan SCC Propagator
+        long count;
+        vector<long> lowlink;
+        vector<long> visit_number;
+        long start_prev_subtree;
+        long end_prev_subtree;
+        long root;
+        long prev_subroot;
+
+        explicit SCCPropagatorData(size_t n) :
+            count(1),
+            lowlink(vector<long>(n, -1)),
+            visit_number(vector<long>(n, -1)),
+            start_prev_subtree(0),
+            end_prev_subtree(0),
+            root(select_root(n)),
+            prev_subroot(root)
+        {
+            lowlink[root] = 0;
+            visit_number[root] = 0;
+        }
+    };
+
+    struct SCCProofData
+    {
+        PosVarDataMap & pos_var_data;
+        ConstraintStateHandle proof_flag_data_handle;
+        ConstraintStateHandle pos_alldiff_data_handle;
+    };
+
+    struct PLine
+    {
+        // Represents a pol line in the proof that we can add terms to.
+        // Maybe this could be generalised (e.g. to other operations) and live in proof.cc?
```

Comment:
Yes please (although `proof.cc` is also too big, and should probably be its own directory rather than living directly in innards)